File: android/app/src/main/kotlin/com/zoonze/zoonze_app/PaymentChannel.kt
package com.zoonze.zoonze_app

import android.app.Activity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject
import payment.sdk.android.payments.PaymentsLauncher
import payment.sdk.android.payments.PaymentsRequest
import payment.sdk.android.payments.PaymentsResult
import payment.sdk.android.savedCard.SavedCardPaymentLauncher
import payment.sdk.android.savedCard.SavedCardPaymentRequest

/**
 * Android half of the `zoonze/payments` MethodChannel — the N-Genius (Network
 * International) card SDK and its Samsung Pay wallet. Contract:
 * docs/backend/payment-contract.md §③.
 *
 * Apple Pay and Samsung Pay are N-Genius *wallets*, not gateways: the `gateway`
 * argument stays `ngenius` for all three and the `wallet` argument selects the
 * SDK entry point. Both ride the same order JSON.
 *
 * Tabby is deliberately not handled here: its SDK ships as a Flutter package
 * (`tabby_flutter_inapp_sdk`), so it lives in Dart and needs no native code on
 * either platform.
 *
 * The launcher is owned by [MainActivity], because `registerForActivityResult`
 * has to run before the activity reaches STARTED and therefore cannot be
 * created on demand when `pay` arrives.
 */
class PaymentChannel {
    companion object {
        private const val CHANNEL = "zoonze/payments"

        /** Order JSON link names — see the SDK's Order model (`_links`). */
        private const val LINK_AUTHORIZATION = "payment-authorization"
        private const val LINK_PAY_PAGE = "payment"

        /** Present on the order JSON when the backend attached a stored card. */
        private const val NODE_SAVED_CARD = "savedCard"

        private const val WALLET_CARD = "card"
        private const val WALLET_APPLE_PAY = "applepay"
        private const val WALLET_SAMSUNG_PAY = "samsungpay"
    }

    private var launcher: PaymentsLauncher? = null
    private var savedCardLauncher: SavedCardPaymentLauncher? = null
    private var samsung: SamsungPaySession? = null

    /**
     * The in-flight Flutter result. Held because the SDK answers on an activity
     * result long after `pay` returns; without it the callback would have
     * nothing to resolve and checkout would hang.
     */
    private var pending: MethodChannel.Result? = null
    private var pendingOrderNumber: String = ""

    /**
     * Ownership token for [pending]. A wallet callback can arrive late — after the
     * customer already backed out and started a card payment — and resolving the
     * wrong (or an already-resolved) result throws. Only the holder of the current
     * token may resolve.
     */
    private var pendingToken: Long = 0
    private var tokenSeq: Long = 0
    private var cardToken: Long = 0

    fun attach(
        engine: FlutterEngine,
        launcher: PaymentsLauncher,
        savedCardLauncher: SavedCardPaymentLauncher,
        activity: Activity,
    ) {
        this.launcher = launcher
        this.savedCardLauncher = savedCardLauncher
        this.samsung = SamsungPaySession(activity)
        MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL)
            .setMethodCallHandler { call, result -> handle(call, result) }
    }

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "pay" -> handlePay(call, result)
            "walletAvailability" -> handleWalletAvailability(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handlePay(call: MethodCall, result: MethodChannel.Result) {
        // Anything that isn't N-Genius is handled in Dart. Reporting "not
        // implemented" lets the Dart layer fall back rather than fail hard.
        if (call.argument<String>("gateway") != "ngenius") {
            result.notImplemented()
            return
        }

        val orderNumber = call.argument<String>("orderNumber").orEmpty()
        when (call.argument<String>("wallet") ?: WALLET_CARD) {
            WALLET_CARD -> startCard(call, result, orderNumber)
            WALLET_SAMSUNG_PAY -> startSamsungPay(call, result, orderNumber)
            // FAILED rather than notImplemented(): "not implemented" means "the
            // native module is missing", which Dart turns into "order awaiting
            // payment". That would be a lie here — this can only be a Dart bug,
            // since walletAvailability already answers applePay=false on Android,
            // and FAILED surfaces it in the payment trace.
            WALLET_APPLE_PAY -> result.success(
                payload("FAILED", orderNumber, raw = "applepay is iOS-only"),
            )
            else -> result.success(
                payload("FAILED", orderNumber, raw = "unsupported wallet"),
            )
        }
    }

    private fun startCard(
        call: MethodCall,
        result: MethodChannel.Result,
        orderNumber: String,
    ) {
        val orderJson = call.argument<String>("orderResponse")

        // Two DIFFERENT order links are required: _links.payment-authorization
        // (the API the SDK authorizes against) and _links.payment (the hosted
        // pay page, which carries the ?code=). Swapping them makes the SDK
        // authorize against the pay page, which answers with an HTML error
        // page rather than a usable response.
        //
        // The order JSON stays authoritative; the channel arguments are a
        // fallback for when order_response is absent.
        val links = parseLinks(orderJson)
        val authorizationUrl = links[LINK_AUTHORIZATION]
            ?: call.argument<String>("authorizationHref")
        val payPageUrl = links[LINK_PAY_PAGE]
            ?: call.argument<String>("payPageHref")

        if (authorizationUrl.isNullOrEmpty() || payPageUrl.isNullOrEmpty()) {
            result.success(
                payload(
                    status = "FAILED",
                    orderNumber = orderNumber,
                    raw = "missing payment links (authorization=" +
                        "${!authorizationUrl.isNullOrEmpty()}, " +
                        "payPage=${!payPageUrl.isNullOrEmpty()})",
                ),
            )
            return
        }

        // A stored card is a different SDK activity: it renders the saved card
        // and recaptures only the CVV. Which one to open is decided here rather
        // than by a channel argument, because the order JSON is authoritative —
        // the backend attaches `savedCard` to the N-Genius order when the quote
        // carried a vault token (docs/backend/payment-contract.md §④), and the
        // SDK reads it from that same order. No CVV crosses the channel; the
        // SDK's own capture screen collects it.
        if (hasSavedCard(orderJson)) {
            val activeSavedCard = savedCardLauncher
            if (activeSavedCard == null) {
                result.success(
                    payload("FAILED", orderNumber, raw = "saved-card launcher unavailable"),
                )
                return
            }
            cardToken = claimPending(result, orderNumber)
            activeSavedCard.launch(
                SavedCardPaymentRequest.builder()
                    .gatewayAuthorizationUrl(authorizationUrl)
                    .payPageUrl(payPageUrl)
                    .build(),
            )
            return
        }

        val activeLauncher = launcher
        if (activeLauncher == null) {
            result.success(
                payload("FAILED", orderNumber, raw = "payment launcher unavailable"),
            )
            return
        }

        cardToken = claimPending(result, orderNumber)

        activeLauncher.launch(
            PaymentsRequest.builder()
                .gatewayAuthorizationUrl(authorizationUrl)
                .payPageUrl(payPageUrl)
                .build(),
        )
    }

    /**
     * Whether the order carries a stored card. Tolerates a malformed body the
     * same way [parseLinks] does — an unreadable order falls back to the normal
     * card form, which is the recoverable direction.
     */
    private fun hasSavedCard(orderJson: String?): Boolean {
        if (orderJson.isNullOrEmpty()) return false
        return try {
            JSONObject(orderJson).optJSONObject(NODE_SAVED_CARD) != null
        } catch (error: Exception) {
            false
        }
    }

    private fun startSamsungPay(
        call: MethodCall,
        result: MethodChannel.Result,
        orderNumber: String,
    ) {
        val serviceId = call.argument<String>("samsungPayServiceId").orEmpty()
        if (serviceId.isBlank()) {
            result.success(
                payload("FAILED", orderNumber, raw = "samsungpay: no service id configured"),
            )
            return
        }
        val session = samsung
        if (session == null) {
            result.success(
                payload("FAILED", orderNumber, raw = "samsungpay: session unavailable"),
            )
            return
        }

        val token = claimPending(result, orderNumber)
        session.start(
            serviceId = serviceId,
            merchantName = call.argument<String>("merchantName").orEmpty().ifEmpty { "Zoonze" },
            orderJson = call.argument<String>("orderResponse"),
        ) { status, raw -> resolvePending(token, status, raw) }
    }

    /**
     * Which wallets this device can pay with. Deliberately independent of
     * [pending]: the checkout list asks this while a payment may be in flight,
     * and answering must never retire someone else's result.
     */
    private fun handleWalletAvailability(call: MethodCall, result: MethodChannel.Result) {
        val serviceId = call.argument<String>("samsungPayServiceId").orEmpty()
        val session = samsung
        if (serviceId.isBlank() || session == null) {
            result.success(mapOf("applePay" to false, "samsungPay" to false))
            return
        }
        session.checkAvailability(serviceId) { available ->
            // applePay is always false on Android — the row is hidden even if the
            // backend offers the method to every device.
            result.success(mapOf("applePay" to false, "samsungPay" to available))
        }
    }

    /**
     * Retires any in-flight payment as cancelled and installs this one. Only one
     * payment can be presented at a time, and resolving a MethodChannel.Result
     * twice throws.
     */
    private fun claimPending(result: MethodChannel.Result, orderNumber: String): Long {
        pending?.success(payload("CANCELLED", pendingOrderNumber, raw = "superseded"))
        pending = result
        pendingOrderNumber = orderNumber
        pendingToken = ++tokenSeq
        return pendingToken
    }

    /** No-ops unless [token] still owns the pending result. */
    private fun resolvePending(token: Long, status: String, raw: String? = null) {
        if (token != pendingToken) return
        val target = pending ?: return
        pending = null
        pendingToken = 0
        val orderNumber = pendingOrderNumber
        pendingOrderNumber = ""
        target.success(payload(status, orderNumber, raw = raw))
    }

    /** Invoked by [MainActivity] when the SDK's activity returns. */
    fun onPaymentResult(result: PaymentsResult) {
        val (status, raw) = when (result) {
            // Authorised now, captured later — the contract treats both as
            // success and the app re-queries Magento for the real state.
            is PaymentsResult.Authorised -> "AUTHORISED" to null
            is PaymentsResult.Success -> "SUCCESS" to null
            is PaymentsResult.PostAuthReview -> "POST_AUTH_REVIEW" to null

            is PaymentsResult.PartialAuthDeclined,
            is PaymentsResult.PartialAuthDeclineFailed,
            -> "DECLINED" to result::class.simpleName

            // Only part of the amount cleared. Passed through unmapped so Dart
            // treats it as failed and routes to CompletePaymentScreen, rather
            // than showing success for an order that isn't fully paid. Matches
            // the iOS module.
            is PaymentsResult.PartiallyAuthorised -> "PARTIALLY_AUTHORISED" to null

            is PaymentsResult.Cancelled -> "CANCELLED" to null
            is PaymentsResult.Failed -> "FAILED" to result.error
        }
        resolvePending(cardToken, status, raw)
    }

    /** Reads `_links` out of the raw order JSON, tolerating a malformed body. */
    private fun parseLinks(orderJson: String?): Map<String, String> {
        if (orderJson.isNullOrEmpty()) return emptyMap()
        return try {
            val links = JSONObject(orderJson).optJSONObject("_links")
                ?: return emptyMap()
            listOf(LINK_AUTHORIZATION, LINK_PAY_PAGE)
                .mapNotNull { name ->
                    links.optJSONObject(name)?.optString("href")
                        ?.takeIf { it.isNotEmpty() }
                        ?.let { name to it }
                }
                .toMap()
        } catch (error: Exception) {
            emptyMap()
        }
    }

    /** Result map shape from the contract. */
    private fun payload(
        status: String,
        orderNumber: String,
        reference: String? = null,
        raw: String? = null,
    ): Map<String, Any> = buildMap {
        put("status", status)
        put("gateway", "ngenius")
        put("orderNumber", orderNumber)
        reference?.let { put("reference", it) }
        raw?.let { put("raw", it) }
    }
}

File: ios/Runner/PaymentChannel.swift
import Flutter
import UIKit

#if canImport(NISdk)
  import NISdk
#endif

/// Native half of the `zoonze/payments` MethodChannel — the N-Genius (Network
/// International) card SDK and its Apple Pay wallet. Contract:
/// docs/backend/payment-contract.md §③.
///
/// The whole NISdk dependency is behind `#if canImport(NISdk)` on purpose. If
/// the pod is missing or fails to install, this still compiles and `pay`
/// answers `FlutterMethodNotImplemented`, which Dart already handles as
/// `MissingPluginException` -> `PaymentGatewayUnavailable` -> "awaiting
/// payment". That's exactly the behaviour before this module existed, so a
/// broken dependency degrades instead of breaking the build or the checkout.
///
/// Tabby is deliberately NOT handled here: its SDK ships as a Flutter package
/// (`tabby_flutter_inapp_sdk`), so it belongs in Dart, not in this module.
final class PaymentChannel: NSObject {
  private static let channelName = "zoonze/payments"

  static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName, binaryMessenger: registrar.messenger())
    let instance = PaymentChannel()
    channel.setMethodCallHandler { call, result in
      instance.handle(call, result: result)
    }
    // Retained by the handler closure above; keep an explicit reference so the
    // delegate isn't deallocated mid-payment while NISdk holds it weakly.
    Self.retained = instance
  }

  private static var retained: PaymentChannel?

  // Both session types are declared inside NISdk's delegate protocols, so the
  // properties are guarded too — the #else path has to compile without the pod.
  #if canImport(NISdk)
    /// Held for the duration of one payment. NISdk keeps only a weak reference
    /// to the delegate, so without this the callback never arrives and the
    /// Flutter result is never fulfilled — the checkout would hang forever.
    private var activeSession: NGeniusSession?

    /// Same reason as `activeSession`.
    private var activeApplePaySession: ApplePaySession?
  #endif

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]

    // Which wallets this device can pay with, so Dart can hide the rows it
    // cannot honour. PassKit is a system framework, so this answers correctly
    // even in a build where the NISdk pod is missing.
    if call.method == "walletAvailability" {
      result([
        "applePay": ApplePayRequestBuilder.canUseApplePay(args),
        // Samsung Pay is Android-only; the row is hidden even if the backend
        // offers the method to every device.
        "samsungPay": false,
      ])
      return
    }

    guard call.method == "pay" else {
      result(FlutterMethodNotImplemented)
      return
    }
    let gateway = args["gateway"] as? String ?? ""

    guard gateway == "ngenius" else {
      // Tabby (and anything else) is handled in Dart. Reporting "not
      // implemented" lets the Dart layer fall back rather than fail hard.
      result(FlutterMethodNotImplemented)
      return
    }

    let orderNumber = args["orderNumber"] as? String ?? ""

    // Apple Pay is an N-Genius *wallet*, not a gateway: same order, same
    // session, different SDK entry point.
    switch args["wallet"] as? String ?? "card" {
    case "card":
      #if canImport(NISdk)
        startNGeniusCard(args: args, result: result)
      #else
        result(FlutterMethodNotImplemented)
      #endif
    case "applepay":
      #if canImport(NISdk)
        startApplePay(args: args, result: result)
      #else
        result(FlutterMethodNotImplemented)
      #endif
    case "samsungpay":
      // FAILED rather than notImplemented(): "not implemented" means "the native
      // module is missing", which Dart turns into "order awaiting payment".
      // That would be a lie — this can only be a Dart bug, since
      // walletAvailability already answers samsungPay=false on iOS.
      result(
        Self.payload(
          status: "FAILED", orderNumber: orderNumber, raw: "samsungpay is Android-only"))
    default:
      result(
        Self.payload(status: "FAILED", orderNumber: orderNumber, raw: "unsupported wallet"))
    }
  }

  #if canImport(NISdk)
    private func startNGeniusCard(args: [String: Any], result: @escaping FlutterResult) {
      let orderNumber = args["orderNumber"] as? String ?? ""

      // iOS drives the SDK from the full order JSON (Android uses the
      // payment-authorization href instead). Both are always sent.
      guard let orderJson = args["orderResponse"] as? String,
        let data = orderJson.data(using: .utf8)
      else {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "orderResponse missing or not a string"))
        return
      }

      let order: OrderResponse
      do {
        order = try OrderResponse.decodeFrom(data: data)
      } catch {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "orderResponse did not decode: \(error)"))
        return
      }

      guard let parent = Self.topViewController() else {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "no view controller to present over"))
        return
      }

      // Render the SDK's own UI in the customer's language.
      if let language = args["language"] as? String, !language.isEmpty {
        NISdk.sharedInstance.setSDKLanguage(language: language)
      }

      let session = NGeniusSession(orderNumber: orderNumber) { [weak self] payload in
        self?.activeSession = nil
        result(payload)
      }
      activeSession = session

      DispatchQueue.main.async {
        NISdk.sharedInstance.showCardPaymentViewWith(
          cardPaymentDelegate: session, overParent: parent, for: order)
      }
    }

    /// Apple Pay on the same N-Genius order the card path uses. The only new
    /// input is the `PKPaymentRequest`, which PassKit needs and the order JSON
    /// does not carry.
    private func startApplePay(args: [String: Any], result: @escaping FlutterResult) {
      let orderNumber = args["orderNumber"] as? String ?? ""

      guard let orderJson = args["orderResponse"] as? String,
        let data = orderJson.data(using: .utf8)
      else {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "orderResponse missing or not a string"))
        return
      }

      let order: OrderResponse
      do {
        order = try OrderResponse.decodeFrom(data: data)
      } catch {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "orderResponse did not decode: \(error)"))
        return
      }

      // Must match `com.apple.developer.in-app-payments` in the entitlements the
      // build was signed with — which, in CI, come from the provisioning profile
      // rather than Runner.entitlements. A blank id here means the build was
      // never configured, and walletAvailability has already hidden the row.
      guard let merchantId = args["applePayMerchantId"] as? String, !merchantId.isEmpty
      else {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "applepay: no merchant identifier"))
        return
      }

      guard let request = ApplePayRequestBuilder.build(args: args, merchantId: merchantId)
      else {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "applepay: could not build PKPaymentRequest"))
        return
      }

      // NISdk presents the PassKit sheet itself, but still wants the presenting
      // controller — same as the card path.
      guard let parent = Self.topViewController() else {
        result(Self.payload(status: "FAILED", orderNumber: orderNumber,
                            raw: "no view controller to present over"))
        return
      }

      let session = ApplePaySession(orderNumber: orderNumber) { [weak self] payload in
        self?.activeApplePaySession = nil
        result(payload)
      }
      activeApplePaySession = session

      // No setSDKLanguage here: the PassKit sheet is rendered by the system in
      // the device locale, and NISdk shows none of its own UI on this path.
      DispatchQueue.main.async {
        NISdk.sharedInstance.initiateApplePayWith(
          applePayDelegate: session, cardPaymentDelegate: session,
          overParent: parent, for: order, with: request)
      }
    }
  #endif

  /// Result map shape from the contract: status / gateway / orderNumber /
  /// reference / raw.
  static func payload(
    status: String, orderNumber: String, reference: String? = nil, raw: String? = nil
  ) -> [String: Any] {
    var map: [String: Any] = [
      "status": status,
      "gateway": "ngenius",
      "orderNumber": orderNumber,
    ]
    if let reference { map["reference"] = reference }
    if let raw { map["raw"] = raw }
    return map
  }

  /// The app is scene-based (see SceneDelegate), so `window` on the app
  /// delegate is not reliable — walk the active scene's key window instead,
  /// then down through anything already presented.
  static func topViewController() -> UIViewController? {
    let window = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .flatMap(\.windows)
      .first { $0.isKeyWindow }
    var top = window?.rootViewController
    while let presented = top?.presentedViewController {
      top = presented
    }
    return top
  }
}

#if canImport(NISdk)
  /// One payment attempt. NISdk reports the outcome through
  /// `CardPaymentDelegate`; this normalises it to the contract's canonical
  /// status string and fulfils the Flutter result exactly once.
  ///
  /// Non-final and internally scoped so `ApplePaySession` can inherit the whole
  /// mapping rather than duplicating (and eventually drifting from) it.
  class NGeniusSession: NSObject, CardPaymentDelegate {
    private let orderNumber: String
    private var complete: (([String: Any]) -> Void)?
    /// Captured from the 3DS/authorization callbacks so a later generic
    /// `PaymentFailed` can be reported as the more specific cause.
    var failureCause: String?

    init(orderNumber: String, complete: @escaping ([String: Any]) -> Void) {
      self.orderNumber = orderNumber
      self.complete = complete
    }

    func finish(_ status: String, raw: String? = nil) {
      // NISdk can fire more than one terminal callback (e.g. a 3DS failure
      // followed by paymentDidComplete). Fulfilling a FlutterResult twice is a
      // hard crash, so only the first one counts.
      guard let complete else { return }
      self.complete = nil
      complete(PaymentChannel.payload(
        status: status, orderNumber: orderNumber, raw: raw))
    }

    func authorizationDidComplete(with status: AuthorizationStatus) {
      if status == .AuthFailed { failureCause = "AUTH_FAILED" }
    }

    func threeDSChallengeDidComplete(with status: ThreeDSStatus) {
      if status == .ThreeDSFailed { failureCause = "THREE_DS_FAILURE" }
    }

    func paymentDidComplete(with status: PaymentStatus) {
      switch status {
      case .PaymentSuccess:
        finish("SUCCESS", raw: status.rawVal)

      // Authorised now, captured asynchronously. The contract treats this as
      // success; the app re-queries Magento for the real order state anyway.
      case .PaymentPostAuthReview:
        finish("POST_AUTH_REVIEW", raw: status.rawVal)

      case .PaymentCancelled:
        finish("CANCELLED", raw: status.rawVal)

      case .PartialAuthDeclined, .PartialAuthDeclineFailed:
        finish("DECLINED", raw: status.rawVal)

      case .PaymentFailed, .InValidRequest:
        // Prefer the specific cause when 3DS or authorization already failed,
        // so the customer sees "declined" rather than a generic error.
        finish(failureCause ?? "FAILED", raw: status.rawVal)

      // Only part of the amount was authorised — not a clean success and not a
      // plain failure. Pass the SDK's own string through: it maps to `failed`
      // in Dart, which routes to CompletePaymentScreen rather than showing a
      // success screen for an order that isn't fully paid.
      case .PartiallyAuthorised:
        finish("PARTIALLY_AUTHORISED", raw: status.rawVal)

      @unknown default:
        finish("FAILED", raw: "unknown PaymentStatus")
      }
    }
  }
#endif
